x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -292,30 +297,41 @@ private void executeCompletionListeners() {
         }
         // we don't need to restore the response headers, they should be included in the current
         // context since we are called by the search action listener.
-        AsyncSearchResponse finalResponse = getResponse();
-        for (Consumer<AsyncSearchResponse> listener : completionListeners.values()) {
-            listener.accept(finalResponse);
-        }
-        completionListeners.clear();
+        getResponse(completionListeners);
```

Comment:
should we copy the completion listeners in the synchronized block to avoid concurrent delete (unregister) ?

Reply:
good idea